// mat-timepicker/esm5/lib/util.js
import { __read } from "tslib";
export function twoDigits(n) {
    return n < 10 ? "0" + n : "" + n;
}
export function addDays(date, days) {
    var result = new Date(date);
    result.setDate(result.getDate() + days);
    return result;
}
export function convertHoursForMode(hour, mode) {
    var isPm = hour >= 12;
    if (mode === '24h') {
        return { hour: hour, isPm: isPm };
    }
    else if (hour === 0 || hour === 12) {
        return { hour: 12, isPm: isPm };
    }
    else if (hour < 12) {
        return { hour: hour, isPm: isPm };
    }
    return { hour: hour - 12, isPm: isPm };
}
function mod(a, b) {
    return a - Math.floor(a / b) * b;
}
export function getShortestAngle(from, to) {
    var difference = to - from;
    return from + mod(difference + 180, 360) - 180;
}
export function isDateInRange(minDate, maxDate, current) {
    var unixCurrentDate = +current;
    return (!minDate || +minDate <= unixCurrentDate) && (!maxDate || unixCurrentDate <= +maxDate);
}
export function isTimeInRange(minDate, maxDate, current) {
    if (minDate instanceof Date) {
        var newMinDate = new Date();
        newMinDate.setHours(minDate.getHours());
        newMinDate.setMinutes(minDate.getMinutes());
        newMinDate.setSeconds(0);
        newMinDate.setMilliseconds(0);
        minDate = newMinDate;
    }
    if (maxDate instanceof Date) {
        var newMaxDate = new Date();
        newMaxDate.setHours(maxDate.getHours());
        newMaxDate.setMinutes(maxDate.getMinutes());
        newMaxDate.setSeconds(0);
        newMaxDate.setMilliseconds(0);
        maxDate = newMaxDate;
    }
    if (current instanceof Date) {
        var newCurrent = new Date();
        newCurrent.setHours(current.getHours());
        newCurrent.setMinutes(current.getMinutes());
        newCurrent.setSeconds(0);
        newCurrent.setMilliseconds(0);
        current = newCurrent;
    }
    var unixCurrentDate = +current;
    return (!minDate || +minDate <= unixCurrentDate) && (!maxDate || unixCurrentDate <= +maxDate);
}
// used when generating the allowed maps
export function isAllowed(hour, minutes, minDate, maxDate, clockMode, selectedMeridiem) {
    if (hour > 24 || hour < 0 || minutes > 60 || minutes < 0) {
        return false;
    }
    if (!minDate && !maxDate) {
        return true;
    }
    if (clockMode === '12h') {
        if (hour === 12 && selectedMeridiem === 'AM') {
            hour = 0;
        }
        if (hour > 12) {
            hour -= 12;
        }
    }
    var checkDate = new Date();
    checkDate.setHours(hour);
    checkDate.setMinutes(minutes);
    checkDate.setSeconds(0);
    checkDate.setMilliseconds(0);
    return isDateInRange(minDate, maxDate, checkDate);
}
// used by the clock component to visually disable the not allowed values
export function getIsAvailabeFn(allowed12HourMap, allowed24HourMap, mode) {
    return function (value, viewType, isPm, h) {
        var isHourCheck = viewType === 'hours';
        var _a = __read(isHourCheck ? [value, null] : [h, value], 2), hour = _a[0], minutes = _a[1];
        if (mode === '12h') {
            if (!allowed12HourMap) {
                return true;
            }
            var meridiem = isPm ? 'pm' : 'am';
            if (isHourCheck) {
                return !!Object.values(allowed12HourMap[meridiem][hour]).find(function (v) { return v === true; });
            }
            return allowed12HourMap[meridiem][hour][minutes];
        }
        if (!allowed24HourMap) {
            return true;
        }
        if (isHourCheck) {
            return !!Object.values(allowed24HourMap[hour]).find(function (v) { return v === true; });
        }
        return allowed24HourMap[hour][minutes];
    };
}
