// mat-timepicker/__ivy_ngcc__/fesm5/mat-timepicker.js
import { __read, __decorate, __param } from 'tslib';
import { NgControl, NgForm, FormGroupDirective } from '@angular/forms';
import { EventEmitter, Inject, Output, Component, Optional, Self, Renderer2, NgZone, ElementRef, Input, HostBinding, HostListener, Directive, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialog, MatDialogModule } from '@angular/material/dialog';
import { MatFormField, MatFormFieldControl } from '@angular/material/form-field';
import { Subject } from 'rxjs';
import { takeUntil, first } from 'rxjs/operators';
import { FocusMonitor } from '@angular/cdk/a11y';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { ErrorStateMatcher } from '@angular/material/core';
import { Platform } from '@angular/cdk/platform';
import { MatButtonModule } from '@angular/material/button';
import { MatToolbarModule } from '@angular/material/toolbar';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { CommonModule } from '@angular/common';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/material/dialog';
import * as ɵngcc2 from '@angular/material/toolbar';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '@angular/material/button';
import * as ɵngcc5 from '@angular/forms';
import * as ɵngcc6 from '@angular/cdk/a11y';
import * as ɵngcc7 from '@angular/cdk/platform';
import * as ɵngcc8 from '@angular/material/form-field';
import * as ɵngcc9 from '@angular/material/core';
import * as ɵngcc10 from 'tslib';

var I18N_0;
if (typeof ngI18nClosureMode !== "undefined" && ngI18nClosureMode) {
    /**
     * @desc Material Timepicker Cancel Button Text
     * @meaning mat-timepicker
     */ 
    var MSG_EXTERNAL_5724442482545923303$$FESM5_MAT_TIMEPICKER_JS_1 = goog.getMsg("{$interpolation}", { "interpolation": "\uFFFD0\uFFFD" });
    I18N_0 = MSG_EXTERNAL_5724442482545923303$$FESM5_MAT_TIMEPICKER_JS_1;
}
else {
    I18N_0 = $localize(ɵngcc10.__makeTemplateObject([":mat-timepicker|Material Timepicker Cancel Button Text\u241F61935f4429dd6d2d9f2c600dae3523b3a22547b0\u241F5724442482545923303:", ":INTERPOLATION:"], [":mat-timepicker|Material Timepicker Cancel Button Text\u241F61935f4429dd6d2d9f2c600dae3523b3a22547b0\u241F5724442482545923303:", ":INTERPOLATION:"]), "\uFFFD0\uFFFD");
}
var I18N_2;
if (typeof ngI18nClosureMode !== "undefined" && ngI18nClosureMode) {
    /**
     * @desc Material Timepicker Ok Button Text
     * @meaning mat-timepicker
     */ 
    var MSG_EXTERNAL_680367210380752514$$FESM5_MAT_TIMEPICKER_JS_3 = goog.getMsg("{$interpolation}", { "interpolation": "\uFFFD0\uFFFD" });
    I18N_2 = MSG_EXTERNAL_680367210380752514$$FESM5_MAT_TIMEPICKER_JS_3;
}
else {
    I18N_2 = $localize(ɵngcc10.__makeTemplateObject([":mat-timepicker|Material Timepicker Ok Button Text\u241F6b9bc53d4851cd32162f49e4cf849fb09ffd4d45\u241F680367210380752514:", ":INTERPOLATION:"], [":mat-timepicker|Material Timepicker Ok Button Text\u241F6b9bc53d4851cd32162f49e4cf849fb09ffd4d45\u241F680367210380752514:", ":INTERPOLATION:"]), "\uFFFD0\uFFFD");
}
function MatTimepickerComponentDialogComponent_ng_container_11_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function MatTimepickerComponentDialogComponent_ng_template_12_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 2);
} }
var _c4 = function (a0) { return { "select": a0 }; };
function MatTimepickerComponentDialogComponent_ng_template_14_Template(rf, ctx) { if (rf & 1) {
    var _r55 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 13);
    ɵngcc0.ɵɵelementStart(1, "span", 14);
    ɵngcc0.ɵɵlistener("click", function MatTimepickerComponentDialogComponent_ng_template_14_Template_span_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r55); var ctx_r54 = ɵngcc0.ɵɵnextContext(); return ctx_r54.setPm(); });
    ɵngcc0.ɵɵtext(2, "PM");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "span", 14);
    ɵngcc0.ɵɵlistener("click", function MatTimepickerComponentDialogComponent_ng_template_14_Template_span_click_3_listener() { ɵngcc0.ɵɵrestoreView(_r55); var ctx_r56 = ɵngcc0.ɵɵnextContext(); return ctx_r56.setAm(); });
    ɵngcc0.ɵɵtext(4, "AM");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r53 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(2, _c4, ctx_r53.isPm));
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(4, _c4, !ctx_r53.isPm));
} }
function ClockComponent_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1, "\u00B7");
    ɵngcc0.ɵɵelementContainerEnd();
} }
function ClockComponent_ng_container_7_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtext(1, "\u00B7");
    ɵngcc0.ɵɵelementContainerEnd();
} }
var _c5 = function (a0, a1) { return { "selected": a0, "disabled": a1 }; };
function ClockComponent_ng_container_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "button", 8);
    ɵngcc0.ɵɵtemplate(2, ClockComponent_ng_container_7_ng_container_2_Template, 2, 0, "ng-container", 6);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var digit_r61 = ctx.$implicit;
    var ctx_r58 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("transform", "translate(" + digit_r61.translateX + "px, " + digit_r61.translateY + "px)");
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(4, _c5, ctx_r58.formattedValue === digit_r61.display || digit_r61.display === 0 && ctx_r58.formattedValue === 0, !ctx_r58.isAvailable(digit_r61.display === 60 ? 0 : digit_r61.display)));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", digit_r61.display % 5 !== 0);
} }
function ClockComponent_ng_container_8_ng_template_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
} if (rf & 2) {
    var digit_r63 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵtextInterpolate(digit_r63.display);
} }
function ClockComponent_ng_container_8_ng_template_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵtext(0);
} if (rf & 2) {
    var digit_r63 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵtextInterpolate(digit_r63.display === 60 ? "00" : digit_r63.display);
} }
function ClockComponent_ng_container_8_ng_container_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainer(0);
} }
function ClockComponent_ng_container_8_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "button", 9);
    ɵngcc0.ɵɵtemplate(2, ClockComponent_ng_container_8_ng_template_2_Template, 1, 1, "ng-template", null, 10, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtemplate(4, ClockComponent_ng_container_8_ng_template_4_Template, 1, 1, "ng-template", null, 11, ɵngcc0.ɵɵtemplateRefExtractor);
    ɵngcc0.ɵɵtemplate(6, ClockComponent_ng_container_8_ng_container_6_Template, 1, 0, "ng-container", 12);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var digit_r63 = ctx.$implicit;
    var _r64 = ɵngcc0.ɵɵreference(3);
    var _r66 = ɵngcc0.ɵɵreference(5);
    var ctx_r59 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("transform", "translate(" + digit_r63.translateX + "px, " + digit_r63.translateY + "px)");
    ɵngcc0.ɵɵproperty("color", ctx_r59.color)("ngClass", ɵngcc0.ɵɵpureFunction2(7, _c5, ctx_r59.formattedValue === digit_r63.display || digit_r63.display === 60 && ctx_r59.formattedValue === 0, !ctx_r59.isAvailable(digit_r63.display === 60 ? 0 : digit_r63.display)));
    ɵngcc0.ɵɵadvance(5);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r59.viewType === "minutes")("ngIfThen", _r66)("ngIfElse", _r64);
} }
function ClockComponent_ng_container_9_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "button", 13);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    var digit_r71 = ctx.$implicit;
    var ctx_r60 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵstyleProp("transform", "translate(" + digit_r71.translateX + "px, " + digit_r71.translateY + "px)");
    ɵngcc0.ɵɵproperty("color", ctx_r60.color)("ngClass", ɵngcc0.ɵɵpureFunction2(5, _c5, ctx_r60.formattedValue === digit_r71.display || digit_r71.display === 24 && ctx_r60.formattedValue === 0, !ctx_r60.isAvailable(digit_r71.display === 24 ? 0 : digit_r71.display)));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", digit_r71.display === 24 ? "00" : digit_r71.display, " ");
} }
var _c6 = function (a0, a1) { return { "small-pointer": a0, "animated-pointer": a1 }; };
var _c7 = function (a0) { return { "outer-dot-odd": a0 }; };
function twoDigits(n) {
    return n < 10 ? "0" + n : "" + n;
}
function addDays(date, days) {
    var result = new Date(date);
    result.setDate(result.getDate() + days);
    return result;
}
function convertHoursForMode(hour, mode) {
    var isPm = hour >= 12;
    if (mode === '24h') {
        return { hour: hour, isPm: isPm };
    }
    else if (hour === 0 || hour === 12) {
        return { hour: 12, isPm: isPm };
    }
    else if (hour < 12) {
        return { hour: hour, isPm: isPm };
    }
    return { hour: hour - 12, isPm: isPm };
}
function mod(a, b) {
    return a - Math.floor(a / b) * b;
}
function getShortestAngle(from, to) {
    var difference = to - from;
    return from + mod(difference + 180, 360) - 180;
}
function isDateInRange(minDate, maxDate, current) {
    var unixCurrentDate = +current;
    return (!minDate || +minDate <= unixCurrentDate) && (!maxDate || unixCurrentDate <= +maxDate);
}
function isTimeInRange(minDate, maxDate, current) {
    if (minDate instanceof Date) {
        var newMinDate = new Date();
        newMinDate.setHours(minDate.getHours());
        newMinDate.setMinutes(minDate.getMinutes());
        newMinDate.setSeconds(0);
        newMinDate.setMilliseconds(0);
        minDate = newMinDate;
    }
    if (maxDate instanceof Date) {
        var newMaxDate = new Date();
        newMaxDate.setHours(maxDate.getHours());
        newMaxDate.setMinutes(maxDate.getMinutes());
        newMaxDate.setSeconds(0);
        newMaxDate.setMilliseconds(0);
        maxDate = newMaxDate;
    }
    if (current instanceof Date) {
        var newCurrent = new Date();
        newCurrent.setHours(current.getHours());
        newCurrent.setMinutes(current.getMinutes());
        newCurrent.setSeconds(0);
        newCurrent.setMilliseconds(0);
        current = newCurrent;
    }
    var unixCurrentDate = +current;
    return (!minDate || +minDate <= unixCurrentDate) && (!maxDate || unixCurrentDate <= +maxDate);
}
// used when generating the allowed maps
function isAllowed(hour, minutes, minDate, maxDate, clockMode, selectedMeridiem) {
    if (hour > 24 || hour < 0 || minutes > 60 || minutes < 0) {
        return false;
    }
    if (!minDate && !maxDate) {
        return true;
    }
    if (clockMode === '12h') {
        if (hour === 12 && selectedMeridiem === 'AM') {
            hour = 0;
        }
        if (hour > 12) {
            hour -= 12;
        }
    }
    var checkDate = new Date();
    checkDate.setHours(hour);
    checkDate.setMinutes(minutes);
    checkDate.setSeconds(0);
    checkDate.setMilliseconds(0);
    return isDateInRange(minDate, maxDate, checkDate);
}
// used by the clock component to visually disable the not allowed values
function getIsAvailabeFn(allowed12HourMap, allowed24HourMap, mode) {
    return function (value, viewType, isPm, h) {
        var isHourCheck = viewType === 'hours';
        var _a = __read(isHourCheck ? [value, null] : [h, value], 2), hour = _a[0], minutes = _a[1];
        if (mode === '12h') {
            if (!allowed12HourMap) {
                return true;
            }
            var meridiem = isPm ? 'pm' : 'am';
            if (isHourCheck) {
                return !!Object.values(allowed12HourMap[meridiem][hour]).find(function (v) { return v === true; });
            }
            return allowed12HourMap[meridiem][hour][minutes];
        }
        if (!allowed24HourMap) {
            return true;
        }
        if (isHourCheck) {
            return !!Object.values(allowed24HourMap[hour]).find(function (v) { return v === true; });
        }
        return allowed24HourMap[hour][minutes];
    };
}

var MatTimepickerComponentDialogComponent = /** @class */ (function () {
    function MatTimepickerComponentDialogComponent(data) {
        this.data = data;
        this.twoDigits = twoDigits;
        this.changeEvent = new EventEmitter();
        this.okClickEvent = new EventEmitter();
        this.cancelClickEvent = new EventEmitter();
        this.allowed24HourMap = null;
        this.allowed12HourMap = null;
        this.invalidSelection = false;
        this.viewType = 'hours';
        this.isPm = false;
        this.skipMinuteAutoSwitch = false;
        this.autoSwitchID = null;
        this.invalidMedianID = null;
        this.hasInvalidMeridiem = false;
        this.editHoursClicked = false;
        this.isClosing = false;
        this.isPm = data.isPm;
        this.bindData(data);
        // keep this always at the bottom
        this.value = data.value;
    }
    Object.defineProperty(MatTimepickerComponentDialogComponent.prototype, "value", {
        get: function () { return this._value; },
        set: function (value) {
            value = value || this.minDate || this.maxDate || new Date();
            this.hours = value.getHours();
            this.minutes = value.getMinutes();
            this._value = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimepickerComponentDialogComponent.prototype, "hours", {
        get: function () { return this._hours; },
        set: function (value) {
            this._hours = value;
            this._formattedHour = convertHoursForMode(this.hours, this.mode).hour;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimepickerComponentDialogComponent.prototype, "formattedHours", {
        get: function () { return this._formattedHour; },
        enumerable: true,
        configurable: true
    });
    MatTimepickerComponentDialogComponent.prototype.bindData = function (data) {
        this.mode = data.mode;
        this.okLabel = data.okLabel;
        this.cancelLabel = data.cancelLabel;
        this.color = data.color;
        this.minDate = data.minDate;
        this.maxDate = data.maxDate;
        this.allowed12HourMap = data.allowed12HourMap;
        this.allowed24HourMap = data.allowed24HourMap;
    };
    MatTimepickerComponentDialogComponent.prototype.ngDoCheck = function () { this.bindData(this.data); };
    MatTimepickerComponentDialogComponent.prototype.handleClockChange = function (_a) {
        var value = _a.value, type = _a.type;
        var is24hoursAutoMeridiemChange = this.mode === '24h' && type === 'hours' && ((this.hours >= 12 && value < 12) || (this.hours < 12 && value >= 12));
        if ((this.hasInvalidMeridiem && this.mode === '12h') || is24hoursAutoMeridiemChange) {
            this.isPm = !this.isPm;
            this.hasInvalidMeridiem = false;
        }
        if ((type && type === 'hours') || (!type && this.viewType === 'hours')) {
            this.hours = value;
        }
        else if ((type && type === 'minutes') || (!type && this.viewType === 'minutes')) {
            this.minutes = value;
        }
        var newValue = new Date();
        var hours = this.isPm ? this.hours < 12 ? this.hours + 12 : this.hours : this.hours === 12 ? 0 : this.hours;
        newValue.setHours(hours);
        newValue.setMinutes(this.minutes);
        newValue.setSeconds(0);
        newValue.setMilliseconds(0);
        this.value = newValue;
        this.changeEvent.emit(newValue);
    };
    MatTimepickerComponentDialogComponent.prototype.clearInvalidMeridiem = function () {
        this.hasInvalidMeridiem = false;
    };
    MatTimepickerComponentDialogComponent.prototype.handleUnavailableSelection = function () {
        clearTimeout(this.autoSwitchID);
    };
    MatTimepickerComponentDialogComponent.prototype.handleClockChangeDone = function (e) {
        var _this = this;
        e.preventDefault(); // prevent mouseUp after touchEnd
        if (this.viewType === 'hours' && !this.skipMinuteAutoSwitch) {
            this.autoSwitchID = setTimeout(function () {
                _this.editMinutes();
                _this.autoSwitchID = null;
            }, 300);
        }
    };
    MatTimepickerComponentDialogComponent.prototype.editHours = function () {
        var _this = this;
        this.viewType = 'hours';
        this.editHoursClicked = true;
        setTimeout(function () { _this.editHoursClicked = false; }, 0);
    };
    MatTimepickerComponentDialogComponent.prototype.editMinutes = function () {
        if (this.hasInvalidMeridiem) {
            this.isPm = !this.isPm;
            this.hasInvalidMeridiem = false;
        }
        this.viewType = 'minutes';
    };
    MatTimepickerComponentDialogComponent.prototype.invalidSelectionHandler = function (value) {
        this.invalidSelection = value;
    };
    MatTimepickerComponentDialogComponent.prototype.invalidMeridiem = function () {
        var _this = this;
        if (this.viewType !== 'minutes' && this.editHoursClicked) {
            if (this.invalidMedianID) {
                return;
            }
            this.invalidMedianID = setTimeout(function () {
                _this.isPm = !_this.isPm;
                _this.hasInvalidMeridiem = false;
            }, 0);
            return;
        }
        this.hasInvalidMeridiem = true;
    };
    MatTimepickerComponentDialogComponent.prototype.meridiemChange = function (hours) {
        var changeData = {
            type: this.viewType,
            value: this.viewType === 'hours' ? hours : this.value.getMinutes()
        };
        this.handleClockChange(changeData);
    };
    MatTimepickerComponentDialogComponent.prototype.setAm = function () {
        if (this.hours >= 12) {
            this.hours = this.hours - 12;
        }
        this.isPm = false;
        this.meridiemChange(this.hours);
    };
    MatTimepickerComponentDialogComponent.prototype.setPm = function () {
        if (this.hours < 12) {
            this.hours = this.hours + 12;
        }
        this.isPm = true;
        this.meridiemChange(this.hours);
    };
    MatTimepickerComponentDialogComponent.prototype.okClickHandler = function () {
        if (this.hasInvalidMeridiem) {
            this.isPm = !this.isPm;
            this.hasInvalidMeridiem = false;
        }
        this.okClickEvent.emit();
    };
    MatTimepickerComponentDialogComponent.prototype.cancelClickHandler = function () {
        this.cancelClickEvent.emit();
    };
    MatTimepickerComponentDialogComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
    ]; };
    __decorate([
        Output()
    ], MatTimepickerComponentDialogComponent.prototype, "changeEvent", void 0);
    __decorate([
        Output()
    ], MatTimepickerComponentDialogComponent.prototype, "okClickEvent", void 0);
    __decorate([
        Output()
    ], MatTimepickerComponentDialogComponent.prototype, "cancelClickEvent", void 0);
    MatTimepickerComponentDialogComponent = __decorate([ __param(0, Inject(MAT_DIALOG_DATA))
    ], MatTimepickerComponentDialogComponent);
MatTimepickerComponentDialogComponent.ɵfac = function MatTimepickerComponentDialogComponent_Factory(t) { return new (t || MatTimepickerComponentDialogComponent)(ɵngcc0.ɵɵdirectiveInject(MAT_DIALOG_DATA)); };
MatTimepickerComponentDialogComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MatTimepickerComponentDialogComponent, selectors: [["mat-timepicker-dialog"]], outputs: { changeEvent: "changeEvent", okClickEvent: "okClickEvent", cancelClickEvent: "cancelClickEvent" }, decls: 23, vars: 28, consts: [[1, "root"], [1, "header", 3, "color"], [1, "placeholder"], [1, "time-frame"], [1, "time", "fixed-font-size", 3, "ngClass", "click"], [1, "fixed-font-size"], [4, "ngIf", "ngIfThen", "ngIfElse"], ["normal", ""], ["ampm", ""], [1, "body"], [3, "allowed12HourMap", "allowed24HourMap", "minDate", "maxDate", "color", "viewType", "mode", "formattedHours", "minutes", "formattedValue", "isPm", "changeEvent", "unavailableSelection", "invalidMeridiem", "mouseup", "clearInvalidMeridiem", "touchend", "invalidSelection"], ["mat-button", "", 3, "color", "click"], ["mat-button", "", 3, "disabled", "color", "click"], [1, "ampm"], [1, "time", 3, "ngClass", "click"]], template: function MatTimepickerComponentDialogComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "mat-dialog-content");
        ɵngcc0.ɵɵelementStart(1, "div", 0);
        ɵngcc0.ɵɵelementStart(2, "mat-toolbar", 1);
        ɵngcc0.ɵɵelement(3, "div", 2);
        ɵngcc0.ɵɵelementStart(4, "div", 3);
        ɵngcc0.ɵɵelementStart(5, "span", 4);
        ɵngcc0.ɵɵlistener("click", function MatTimepickerComponentDialogComponent_Template_span_click_5_listener() { return ctx.editHours(); });
        ɵngcc0.ɵɵtext(6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(7, "span", 5);
        ɵngcc0.ɵɵtext(8, ":");
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(9, "span", 4);
        ɵngcc0.ɵɵlistener("click", function MatTimepickerComponentDialogComponent_Template_span_click_9_listener() { return ctx.editMinutes(); });
        ɵngcc0.ɵɵtext(10);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(11, MatTimepickerComponentDialogComponent_ng_container_11_Template, 1, 0, "ng-container", 6);
        ɵngcc0.ɵɵtemplate(12, MatTimepickerComponentDialogComponent_ng_template_12_Template, 1, 0, "ng-template", null, 7, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵtemplate(14, MatTimepickerComponentDialogComponent_ng_template_14_Template, 5, 6, "ng-template", null, 8, ɵngcc0.ɵɵtemplateRefExtractor);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(16, "div", 9);
        ɵngcc0.ɵɵelementStart(17, "mat-clock", 10);
        ɵngcc0.ɵɵlistener("changeEvent", function MatTimepickerComponentDialogComponent_Template_mat_clock_changeEvent_17_listener($event) { return ctx.handleClockChange($event); })("unavailableSelection", function MatTimepickerComponentDialogComponent_Template_mat_clock_unavailableSelection_17_listener() { return ctx.handleUnavailableSelection(); })("invalidMeridiem", function MatTimepickerComponentDialogComponent_Template_mat_clock_invalidMeridiem_17_listener() { return ctx.invalidMeridiem(); })("mouseup", function MatTimepickerComponentDialogComponent_Template_mat_clock_mouseup_17_listener($event) { return ctx.handleClockChangeDone($event); })("clearInvalidMeridiem", function MatTimepickerComponentDialogComponent_Template_mat_clock_clearInvalidMeridiem_17_listener() { return ctx.clearInvalidMeridiem(); })("touchend", function MatTimepickerComponentDialogComponent_Template_mat_clock_touchend_17_listener($event) { return ctx.handleClockChangeDone($event); })("invalidSelection", function MatTimepickerComponentDialogComponent_Template_mat_clock_invalidSelection_17_listener($event) { return ctx.invalidSelectionHandler($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(18, "mat-dialog-actions");
        ɵngcc0.ɵɵelementStart(19, "button", 11);
        ɵngcc0.ɵɵlistener("click", function MatTimepickerComponentDialogComponent_Template_button_click_19_listener() { return ctx.cancelClickHandler(); });
        ɵngcc0.ɵɵi18n(20, I18N_0);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(21, "button", 12);
        ɵngcc0.ɵɵlistener("click", function MatTimepickerComponentDialogComponent_Template_button_click_21_listener() { return ctx.okClickHandler(); });
        ɵngcc0.ɵɵi18n(22, I18N_2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        var _r50 = ɵngcc0.ɵɵreference(13);
        var _r52 = ɵngcc0.ɵɵreference(15);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("color", ctx.color);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(24, _c4, ctx.viewType === "hours" && "active"));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.twoDigits(ctx.formattedHours), " ");
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(26, _c4, ctx.viewType === "minutes" && "active"));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵtextInterpolate1(" ", ctx.twoDigits(ctx.minutes), " ");
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.mode === "12h")("ngIfThen", _r52)("ngIfElse", _r50);
        ɵngcc0.ɵɵadvance(6);
        ɵngcc0.ɵɵproperty("allowed12HourMap", ctx.allowed12HourMap)("allowed24HourMap", ctx.allowed24HourMap)("minDate", ctx.minDate)("maxDate", ctx.maxDate)("color", ctx.color)("viewType", ctx.viewType)("mode", ctx.mode)("formattedHours", ctx.formattedHours)("minutes", ctx.minutes)("formattedValue", ctx.viewType === "minutes" ? ctx.minutes : ctx.formattedHours)("isPm", ctx.isPm);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("color", ctx.color);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵi18nExp(ctx.cancelLabel);
        ɵngcc0.ɵɵi18nApply(20);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("disabled", ctx.invalidSelection)("color", ctx.color);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵi18nExp(ctx.okLabel);
        ɵngcc0.ɵɵi18nApply(22);
    } }, directives: function () { return [ɵngcc1.MatDialogContent, ɵngcc2.MatToolbar, ɵngcc3.NgClass, ɵngcc3.NgIf, ClockComponent, ɵngcc1.MatDialogActions, ɵngcc4.MatButton]; }, styles: ["mat-dialog-content[_ngcontent-%COMP%]{min-height:395px;padding:0;margin-top:-24px;overflow:hidden}mat-dialog-actions[_ngcontent-%COMP%]{justify-content:flex-end;margin-right:-8px;margin-left:-8px}.root[_ngcontent-%COMP%]{min-width:282px}.header[_ngcontent-%COMP%]{border-top-left-radius:2px;border-top-right-radius:2px;padding:20px 0;line-height:58px;font-size:58px;display:flex;justify-content:center;align-items:center;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;height:98px}.header[_ngcontent-%COMP%]   .fixed-font-size[_ngcontent-%COMP%]{font-size:58px}.header[_ngcontent-%COMP%]   .time-frame[_ngcontent-%COMP%]{height:60px}.time[_ngcontent-%COMP%]{transition:.2s ease-out;cursor:pointer}.time[_ngcontent-%COMP%]:not(.select){opacity:.6}.placeholder[_ngcontent-%COMP%]{flex:1}.ampm[_ngcontent-%COMP%]{display:flex;flex-direction:column-reverse;flex:1;font-size:14px;line-height:20px;margin-left:16px;font-weight:700px}.select[_ngcontent-%COMP%]{color:#fff}.body[_ngcontent-%COMP%]{padding:24px 16px 20px;display:flex;justify-content:center}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatTimepickerComponentDialogComponent, [{
        type: Component,
        args: [{
                selector: 'mat-timepicker-dialog',
                template: "<mat-dialog-content>\n  <div class=\"root\">\n    <mat-toolbar [color]=\"color\" class=\"header\">\n      <div class=\"placeholder\"></div>\n      <div class=\"time-frame\">\n        <span class=\"time fixed-font-size\" [ngClass]=\"{'select': this.viewType === 'hours' && 'active' }\"\n          (click)=\"editHours()\">\n          {{ twoDigits(formattedHours) }}\n        </span>\n        <span class=\"fixed-font-size\">:</span>\n        <span class=\"time fixed-font-size\" [ngClass]=\"{ 'select': this.viewType === 'minutes' && 'active' }\"\n          (click)=\"editMinutes()\">\n          {{ twoDigits(minutes) }}\n        </span>\n      </div>\n      <ng-container *ngIf=\"mode === '12h' then ampm else normal\"></ng-container>\n      <ng-template #normal>\n        <div class=\"placeholder\"></div>\n      </ng-template>\n      <ng-template #ampm>\n        <div class=\"ampm\">\n          <span class=\"time\" [ngClass]=\"{ 'select': isPm }\" (click)=\"setPm()\">PM</span>\n          <span class=\"time\" [ngClass]=\"{ 'select': !isPm }\" (click)=\"setAm()\">AM</span>\n        </div>\n      </ng-template>\n    </mat-toolbar>\n    <div class=\"body\">\n      <mat-clock [allowed12HourMap]=\"allowed12HourMap\" [allowed24HourMap]=\"allowed24HourMap\" [minDate]=\"minDate\"\n        [maxDate]=\"maxDate\" [color]=\"color\" [viewType]=\"viewType\" [mode]=\"mode\" [formattedHours]=\"formattedHours\"\n        [minutes]=\"minutes\" (changeEvent)=\"handleClockChange($event)\"\n        (unavailableSelection)=\"handleUnavailableSelection()\"\n        [formattedValue]=\"viewType === 'minutes' ? minutes : formattedHours\" [isPm]=\"isPm\"\n        (invalidMeridiem)=\"invalidMeridiem()\" (mouseup)=\"handleClockChangeDone($event)\"\n        (clearInvalidMeridiem)=\"clearInvalidMeridiem()\" (touchend)=\"handleClockChangeDone($event)\"\n        (invalidSelection)=\"invalidSelectionHandler($event)\"></mat-clock>\n    </div>\n  </div>\n</mat-dialog-content>\n<mat-dialog-actions>\n  <button mat-button i18n=\"mat-timepicker|Material Timepicker Cancel Button Text\" [color]=\"color\"\n    (click)=\"cancelClickHandler()\">{{cancelLabel}}</button>\n  <button mat-button i18n=\"mat-timepicker|Material Timepicker Ok Button Text\" [disabled]=\"invalidSelection\"\n    [color]=\"color\" (click)=\"okClickHandler()\">{{okLabel}}</button>\n</mat-dialog-actions>",
                styles: ["mat-dialog-content{min-height:395px;padding:0;margin-top:-24px;overflow:hidden}mat-dialog-actions{justify-content:flex-end;margin-right:-8px;margin-left:-8px}.root{min-width:282px}.header{border-top-left-radius:2px;border-top-right-radius:2px;padding:20px 0;line-height:58px;font-size:58px;display:flex;justify-content:center;align-items:center;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;height:98px}.header .fixed-font-size{font-size:58px}.header .time-frame{height:60px}.time{transition:.2s ease-out;cursor:pointer}.time:not(.select){opacity:.6}.placeholder{flex:1}.ampm{display:flex;flex-direction:column-reverse;flex:1;font-size:14px;line-height:20px;margin-left:16px;font-weight:700px}.select{color:#fff}.body{padding:24px 16px 20px;display:flex;justify-content:center}"]
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [MAT_DIALOG_DATA]
            }] }]; }, { changeEvent: [{
            type: Output
        }], okClickEvent: [{
            type: Output
        }], cancelClickEvent: [{
            type: Output
        }] }); })();
    return MatTimepickerComponentDialogComponent;
}());

var MatTimepickerDirective = /** @class */ (function () {
    function MatTimepickerDirective(ngControl, dialog, renderer, zone, fm, elRef, ngZone, 
    // tslint:disable-next-line:variable-name
    _platform, 
    // tslint:disable-next-line:variable-name
    _parentForm, 
    // tslint:disable-next-line:variable-name
    _matFormFiled, 
    // tslint:disable-next-line:variable-name
    _parentFormGroup, 
    // tslint:disable-next-line:variable-name
    _defaultErrorStateMatcher) {
        var _this = this;
        this.ngControl = ngControl;
        this.dialog = dialog;
        this.renderer = renderer;
        this.zone = zone;
        this.fm = fm;
        this.elRef = elRef;
        this.ngZone = ngZone;
        this._platform = _platform;
        this._parentForm = _parentForm;
        this._matFormFiled = _matFormFiled;
        this._parentFormGroup = _parentFormGroup;
        // tslint:disable-next-line:variable-name
        this._errorState = false;
        // tslint:disable-next-line:variable-name
        this._disabled = false;
        // tslint:disable-next-line:variable-name
        this._readonly = false;
        this.isAlive = new Subject();
        this.stateChanges = new Subject();
        // tslint:disable-next-line:variable-name
        this._uid = "mat-input-" + MatTimepickerDirective_1.nextId++;
        this.describedBy = '';
        // tslint:disable-next-line:variable-name
        this._required = false;
        this.focused = false;
        this.allowed24HourMap = null;
        this.allowed12HourMap = null;
        this.isInputFocused = false;
        /** Override the label of the ok button. */
        this.okLabel = 'Ok';
        /** Override the label of the cancel button. */
        this.cancelLabel = 'Cancel';
        /** Sets the clock mode, 12-hour or 24-hour clocks are supported. */
        this.mode = '24h';
        this.color = 'primary';
        this.disableDialogOpenOnClick = false;
        this.strict = true;
        this.controlType = 'angular-material-timepicker';
        this.listeners = [];
        // tslint:disable-next-line:variable-name
        this._skipValueChangeEmission = true;
        this.combination = [];
        this.timeChange = new EventEmitter();
        this.invalidInput = new EventEmitter();
        this.clickHandler = function (e) {
            if ((_this.modalRef && _this.modalRef.componentInstance.isClosing) || _this.disabled || _this.disableDialogOpenOnClick) {
                return;
            }
            if (!_this.modalRef && !_this.disableDialogOpenOnClick) {
                _this.showDialog();
            }
        };
        this.handleChange = function (newValue) {
            if (!(newValue instanceof Date)) {
                return;
            }
            var v = _this.value instanceof Date ? new Date(_this.value.getTime()) : new Date();
            v.setHours(newValue.getHours());
            v.setMinutes(newValue.getMinutes());
            v.setSeconds(0);
            v.setMilliseconds(0);
            _this.currentValue = v;
        };
        this.handleOk = function () {
            if (_this.onChangeFn) {
                _this.onChangeFn(_this.currentValue);
            }
            _this.value = _this.currentValue;
            _this.modalRef.close();
        };
        this.handleCancel = function () {
            _this.modalRef.close();
        };
        this.id = this.id;
        this.errorStateMatcher = _defaultErrorStateMatcher;
        if (this.ngControl != null) {
            this.ngControl.valueAccessor = this;
        }
        if (_platform.IOS) {
            ngZone.runOutsideAngular(function () {
                elRef.nativeElement.addEventListener('keyup', function (event) {
                    var el = event.target;
                    if (!el.value && !el.selectionStart && !el.selectionEnd) {
                        // Note: Just setting `0, 0` doesn't fix the issue. Setting
                        // `1, 1` fixes it for the first time that you type text and
                        // then hold delete. Toggling to `1, 1` and then back to
                        // `0, 0` seems to completely fix it.
                        el.setSelectionRange(1, 1);
                        el.setSelectionRange(0, 0);
                    }
                });
            });
        }
        this._isServer = !this._platform.isBrowser;
    }
    MatTimepickerDirective_1 = MatTimepickerDirective;
    Object.defineProperty(MatTimepickerDirective.prototype, "errorState", {
        get: function () {
            var oldState = this._errorState;
            var parent = this._parentFormGroup || this._parentForm;
            var control = this.ngControl ? this.ngControl.control : null;
            var newState = this.errorStateMatcher ? this.errorStateMatcher.isErrorState(control, parent) : oldState;
            if (newState !== oldState) {
                this._errorState = newState;
                this.stateChanges.next();
            }
            return newState;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimepickerDirective.prototype, "disabled", {
        get: function () {
            if (this.ngControl && this.ngControl.disabled !== null) {
                return this.ngControl.disabled;
            }
            return this._disabled;
        },
        set: function (value) {
            this._disabled = coerceBooleanProperty(value);
            // Browsers may not fire the blur event if the input is disabled too quickly.
            // Reset from here to ensure that the element doesn't become stuck.
            if (this.focused) {
                this.focused = false;
                this.stateChanges.next();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimepickerDirective.prototype, "id", {
        get: function () { return this._id; },
        set: function (value) { this._id = value || this._uid; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimepickerDirective.prototype, "readonly", {
        get: function () { return this._readonly; },
        set: function (value) { this._readonly = coerceBooleanProperty(value); },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimepickerDirective.prototype, "shouldLabelFloat", {
        get: function () { return this.focused || !this.empty; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimepickerDirective.prototype, "required", {
        get: function () {
            return this._required;
        },
        set: function (req) {
            this._required = coerceBooleanProperty(req);
            this.stateChanges.next();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimepickerDirective.prototype, "placeholder", {
        get: function () {
            return this._placeholder;
        },
        set: function (plh) {
            this._placeholder = plh;
            this.stateChanges.next();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimepickerDirective.prototype, "value", {
        get: function () { return this._value; },
        set: function (value) {
            if (value === this._value) {
                return;
            }
            this._value = value;
            if (!value) {
                this._formattedValueString = null;
                this.setInputElementValue(value);
                this.currentValue = value;
                return;
            }
            var _a = convertHoursForMode(value.getHours(), this.mode), hour = _a.hour, isPm = _a.isPm;
            this._isPm = isPm;
            this._formattedValueString = this.mode === '12h' ?
                hour + ":" + twoDigits(value.getMinutes()) + " " + (isPm ? 'pm' : 'am') :
                twoDigits(value.getHours()) + ":" + twoDigits(value.getMinutes());
            if (!this.isInputFocused) {
                this.setInputElementValue(this.formattedValueString);
            }
            this.currentValue = value;
            this.stateChanges.next();
            if (this._skipValueChangeEmission) {
                return;
            }
            this.timeChange.emit(this.currentValue);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimepickerDirective.prototype, "isPm", {
        get: function () { return this._isPm; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimepickerDirective.prototype, "empty", {
        get: function () {
            return !(this.currentValue instanceof Date);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimepickerDirective.prototype, "formattedValueString", {
        get: function () { return this._formattedValueString; },
        enumerable: true,
        configurable: true
    });
    MatTimepickerDirective.prototype.inputHandler = function () {
        var _a, _b;
        var value = this.elRef.nativeElement.value;
        var length = value.length;
        if (length === 0) {
            this.writeValue(null, true);
            if (this.onChangeFn) {
                this.onChangeFn(null);
            }
            return;
        }
        var meridiemResult = value.match(/am|pm/i);
        var meridiem = null;
        if (meridiemResult) {
            value = value.replace(meridiemResult[0], '');
            _a = __read(meridiemResult, 1), meridiem = _a[0];
        }
        var valueHasColumn = value.includes(':');
        var _c = __read(length === 1 ? [value, 0] :
            length === 2 && !valueHasColumn ? [value, 0] : valueHasColumn ? value.split(':') : value.split(/(\d\d)/).filter(function (v) { return v; }), 2), hours = _c[0], minutes = _c[1];
        hours = +hours;
        if (/\s/.test(minutes)) {
            var other = void 0;
            _b = __read(minutes.split(/\s/), 2), minutes = _b[0], other = _b[1];
            if (other === 'pm' && !isNaN(hours) && hours < 12) {
                hours += 12;
            }
        }
        minutes = +minutes;
        if (isNaN(hours) || isNaN(minutes)) {
            this.writeValue(null, true);
            return;
        }
        if (hours < 12 && meridiem && meridiem.toLowerCase() === 'pm') {
            hours += 12;
        }
        else if (hours > 12 && meridiem && meridiem.toLowerCase() === 'am') {
            hours -= 12;
        }
        if (this.mode === '12h' && +hours < 1) {
            hours = '1';
        }
        else {
            if (+hours > 24) {
                hours = '24';
            }
            else if (+hours < 0) {
                hours = '0';
            }
        }
        if (+minutes > 59) {
            minutes = '59';
        }
        else if (+minutes < 0) {
            minutes = '0';
        }
        var d = this.value ? new Date(this.value.getTime()) : new Date();
        d.setHours(+hours);
        d.setMinutes(+minutes);
        d.setSeconds(0);
        d.setMilliseconds(0);
        var isValueInRange = isDateInRange(this.minDate, this.maxDate, d);
        if (!isValueInRange) {
            this.invalidInput.emit();
        }
        this.writeValue(d, true);
        if (this.onChangeFn) {
            this.onChangeFn(d);
        }
    };
    MatTimepickerDirective.prototype.keydownHandler = function (event) {
        if (event.metaKey || event.ctrlKey || event.altKey) {
            this.combination = this.combination.concat(event.code);
            return;
        }
        if (!/^[0-9a-zA-Z\s]{0,1}$/.test(event.key)) {
            return;
        }
        var target = event.target;
        var tValue = target.value;
        var value = "" + tValue.slice(0, target.selectionStart) + event.key + tValue.slice(target.selectionEnd);
        if (value.match(this.pattern) || this.combination.length > 0) {
            return true;
        }
        event.preventDefault();
        event.stopImmediatePropagation();
    };
    MatTimepickerDirective.prototype.keyupHandler = function (event) {
        this.combination = this.combination.filter(function (v) { return v !== event.code; });
    };
    MatTimepickerDirective.prototype.focusHandler = function () {
        this.isInputFocused = true;
    };
    MatTimepickerDirective.prototype.focusoutHandler = function () {
        this.isInputFocused = false;
        this.setInputElementValue(this.formattedValueString);
        if (this.onTouchedFn && !this.modalRef) {
            this.onTouchedFn();
        }
    };
    MatTimepickerDirective.prototype.setDescribedByIds = function (ids) {
        this.describedBy = ids.join(' ');
    };
    MatTimepickerDirective.prototype.onContainerClick = function (event) {
        if (event.target.tagName.toLowerCase() !== 'input') {
            this.elRef.nativeElement.focus();
        }
    };
    MatTimepickerDirective.prototype.setInputElementValue = function (value) {
        var _this = this;
        Promise.resolve().then(function () {
            _this.zone.runOutsideAngular(function () {
                _this.renderer.setProperty(_this.elRef.nativeElement, 'value', value);
            });
        });
    };
    MatTimepickerDirective.prototype.validate = function () {
        var isValueInRange = this.strict ?
            isDateInRange(this.minDate, this.maxDate, this.currentValue) :
            isTimeInRange(this.minDate, this.maxDate, this.currentValue);
        return isValueInRange ? null : { dateRange: true };
    };
    MatTimepickerDirective.prototype.ngAfterViewInit = function () {
        this.listeners.push(this.renderer.listen(this._matFormFiled ? this._matFormFiled._elementRef.nativeElement : this.elRef.nativeElement, 'click', this.clickHandler));
    };
    MatTimepickerDirective.prototype.ngOnInit = function () {
        var _this = this;
        if (this._platform.isBrowser) {
            this.fm.monitor(this.elRef.nativeElement, true).subscribe(function (origin) {
                _this.focused = !!origin;
                _this.stateChanges.next();
            });
        }
        if (!this.value) {
            var hasMaxDate = !!this.maxDate;
            var hasMinDate = !!this.minDate;
            if (hasMinDate || hasMaxDate) {
                if (hasMinDate) {
                    this.minDate.setSeconds(0);
                    this.minDate.setMilliseconds(0);
                }
                if (hasMaxDate) {
                    this.maxDate.setSeconds(0);
                    this.maxDate.setMilliseconds(0);
                }
                Promise.resolve().then(function () { return _this.generateAllowedMap(); });
                if (!this.ngControl._rawValidators.find(function (v) { return v === _this; })) {
                    this.ngControl._rawValidators.push(this);
                }
            }
        }
        this._skipValueChangeEmission = false;
    };
    MatTimepickerDirective.prototype.generateAllowedMap = function () {
        var isStrictMode = this.strict && this.value instanceof Date;
        if (this.mode === '24h') {
            this.allowed24HourMap = {};
            for (var h = 0; h < 24; h++) {
                for (var m = 0; m < 60; m++) {
                    var hourMap = this.allowed24HourMap[h] || {};
                    if (isStrictMode) {
                        var currentDate = new Date(this.value.getTime());
                        currentDate.setHours(h);
                        currentDate.setMinutes(m);
                        currentDate.setSeconds(0);
                        currentDate.setMilliseconds(0);
                        hourMap[m] = isDateInRange(this.minDate, this.maxDate, currentDate);
                    }
                    else {
                        hourMap[m] = isAllowed(h, m, this.minDate, this.maxDate, '24h');
                    }
                    this.allowed24HourMap[h] = hourMap;
                }
            }
        }
        else {
            this.allowed12HourMap = { am: {}, pm: {} };
            for (var h = 0; h < 24; h++) {
                var meridiem = h < 12 ? 'am' : 'pm';
                for (var m = 0; m < 60; m++) {
                    var hour = (h > 12 ? h - 12 : h === 0 ? 12 : h);
                    var hourMap = this.allowed12HourMap[meridiem][hour] || {};
                    if (isStrictMode) {
                        var currentDate = new Date(this.value.getTime());
                        currentDate.setHours(h);
                        currentDate.setMinutes(m);
                        currentDate.setSeconds(0);
                        currentDate.setMilliseconds(0);
                        hourMap[m] = isDateInRange(this.minDate, this.maxDate, currentDate);
                    }
                    else {
                        hourMap[m] = isAllowed(h, m, this.minDate, this.maxDate, '24h');
                    }
                    this.allowed12HourMap[meridiem][hour] = hourMap;
                }
            }
        }
    };
    MatTimepickerDirective.prototype.ngOnChanges = function (simpleChanges) {
        this.pattern = this.mode === '24h' ? /^[0-9]{1,2}:?([0-9]{1,2})?$/ : /^[0-9]{1,2}:?([0-9]{1,2})?\s?(a|p)?m?$/;
        if ((simpleChanges.minDate && !simpleChanges.minDate.isFirstChange &&
            +simpleChanges.minDate.currentValue !== simpleChanges.minDate.previousValue) ||
            (simpleChanges.maxDate && !simpleChanges.maxDate.isFirstChange &&
                +simpleChanges.maxDate.currentValue !== simpleChanges.maxDate.previousValue) ||
            (simpleChanges.disableLimitBase && !simpleChanges.disableLimitBase.isFirstChange &&
                +simpleChanges.disableLimitBase.currentValue !== simpleChanges.disableLimitBase.previousValue)) {
            this.generateAllowedMap();
        }
        if (!this.modalRef || !this.modalRef.componentInstance) {
            return;
        }
        this.modalRef.componentInstance.data = {
            mode: this.mode,
            value: this.currentValue,
            okLabel: this.okLabel,
            cancelLabel: this.cancelLabel,
            color: this.color,
            isPm: this.isPm,
            minDate: this.minDate,
            maxDate: this.maxDate,
            allowed12HourMap: this.allowed12HourMap,
            allowed24HourMap: this.allowed24HourMap
        };
    };
    MatTimepickerDirective.prototype.checkValidity = function (value) {
        if (!value) {
            return false;
        }
        var hour = value.getHours();
        var minutes = value.getMinutes();
        var meridiem = this.isPm ? 'PM' : 'AM';
        return isAllowed(hour, minutes, this.minDate, this.maxDate, this.mode, meridiem);
    };
    MatTimepickerDirective.prototype.writeValue = function (value, isInnerCall) {
        var _this = this;
        if (isInnerCall === void 0) { isInnerCall = false; }
        if (!isInnerCall) {
            this._skipValueChangeEmission = true;
            Promise.resolve().then(function () { return _this._skipValueChangeEmission = false; });
        }
        if (value) {
            value.setSeconds(0);
            value.setMilliseconds(0);
        }
        if (+this.value !== +value) {
            this.value = value;
        }
    };
    MatTimepickerDirective.prototype.registerOnChange = function (fn) {
        this.onChangeFn = fn;
    };
    MatTimepickerDirective.prototype.registerOnTouched = function (fn) {
        this.onTouchedFn = fn;
    };
    MatTimepickerDirective.prototype.setDisabledState = function (isDisabled) {
        this.disabled = isDisabled;
    };
    MatTimepickerDirective.prototype.showDialog = function () {
        var _this = this;
        if (this.disabled) {
            return;
        }
        this.isInputFocused = false;
        this.modalRef = this.dialog.open(MatTimepickerComponentDialogComponent, {
            autoFocus: false,
            data: {
                mode: this.mode,
                value: this.currentValue,
                okLabel: this.okLabel,
                cancelLabel: this.cancelLabel,
                color: this.color,
                isPm: this.isPm,
                minDate: this.minDate,
                maxDate: this.maxDate,
                allowed12HourMap: this.allowed12HourMap,
                allowed24HourMap: this.allowed24HourMap
            }
        });
        var instance = this.modalRef.componentInstance;
        instance.changeEvent.pipe(takeUntil(this.isAlive)).subscribe(this.handleChange);
        instance.okClickEvent.pipe(takeUntil(this.isAlive)).subscribe(this.handleOk);
        instance.cancelClickEvent.pipe(takeUntil(this.isAlive)).subscribe(this.handleCancel);
        this.modalRef.beforeClosed().pipe(first()).subscribe(function () { return instance.isClosing = true; });
        this.modalRef.afterClosed().pipe(first()).subscribe(function () {
            if (_this.onTouchedFn) {
                _this.onTouchedFn();
            }
            _this.modalRef = null;
            _this.elRef.nativeElement.focus();
        });
        this.currentValue = this.value;
    };
    MatTimepickerDirective.prototype.ngOnDestroy = function () {
        this.isAlive.next();
        this.isAlive.complete();
        this.stateChanges.complete();
        if (this._platform.isBrowser) {
            this.fm.stopMonitoring(this.elRef.nativeElement);
        }
        this.listeners.forEach(function (l) { return l(); });
    };
    var MatTimepickerDirective_1;
    MatTimepickerDirective.nextId = 0;
    MatTimepickerDirective.ctorParameters = function () { return [
        { type: NgControl, decorators: [{ type: Optional }, { type: Self }] },
        { type: MatDialog },
        { type: Renderer2 },
        { type: NgZone },
        { type: FocusMonitor },
        { type: ElementRef },
        { type: NgZone },
        { type: Platform },
        { type: NgForm, decorators: [{ type: Optional }] },
        { type: MatFormField, decorators: [{ type: Optional }] },
        { type: FormGroupDirective, decorators: [{ type: Optional }] },
        { type: ErrorStateMatcher }
    ]; };
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "disabled", null);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "id", null);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "readonly", null);
    __decorate([
        HostBinding('class.floating')
    ], MatTimepickerDirective.prototype, "shouldLabelFloat", null);
    __decorate([
        HostBinding('attr.aria-describedby')
    ], MatTimepickerDirective.prototype, "describedBy", void 0);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "errorStateMatcher", void 0);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "required", null);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "placeholder", null);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "okLabel", void 0);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "cancelLabel", void 0);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "mode", void 0);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "color", void 0);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "disableDialogOpenOnClick", void 0);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "strict", void 0);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "minDate", void 0);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "maxDate", void 0);
    __decorate([
        Input()
    ], MatTimepickerDirective.prototype, "value", null);
    __decorate([
        Output()
    ], MatTimepickerDirective.prototype, "timeChange", void 0);
    __decorate([
        Output()
    ], MatTimepickerDirective.prototype, "invalidInput", void 0);
    __decorate([
        HostListener('input')
    ], MatTimepickerDirective.prototype, "inputHandler", null);
    __decorate([
        HostListener('keydown', ['$event'])
    ], MatTimepickerDirective.prototype, "keydownHandler", null);
    __decorate([
        HostListener('keyup', ['$event'])
    ], MatTimepickerDirective.prototype, "keyupHandler", null);
    __decorate([
        HostListener('focus')
    ], MatTimepickerDirective.prototype, "focusHandler", null);
    __decorate([
        HostListener('focusout')
    ], MatTimepickerDirective.prototype, "focusoutHandler", null);
    MatTimepickerDirective = MatTimepickerDirective_1 = __decorate([ __param(0, Optional()), __param(0, Self()),
        __param(8, Optional()),
        __param(9, Optional()),
        __param(10, Optional())
    ], MatTimepickerDirective);
MatTimepickerDirective.ɵfac = function MatTimepickerDirective_Factory(t) { return new (t || MatTimepickerDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc5.NgControl, 10), ɵngcc0.ɵɵdirectiveInject(ɵngcc1.MatDialog), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone), ɵngcc0.ɵɵdirectiveInject(ɵngcc6.FocusMonitor), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.NgZone), ɵngcc0.ɵɵdirectiveInject(ɵngcc7.Platform), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.NgForm, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc8.MatFormField, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc5.FormGroupDirective, 8), ɵngcc0.ɵɵdirectiveInject(ɵngcc9.ErrorStateMatcher)); };
MatTimepickerDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: MatTimepickerDirective, selectors: [["input", "matTimepicker", ""]], hostAttrs: [1, "mat-input-element", "mat-form-field-autofill-control"], hostVars: 12, hostBindings: function MatTimepickerDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("input", function MatTimepickerDirective_input_HostBindingHandler() { return ctx.inputHandler(); })("keydown", function MatTimepickerDirective_keydown_HostBindingHandler($event) { return ctx.keydownHandler($event); })("keyup", function MatTimepickerDirective_keyup_HostBindingHandler($event) { return ctx.keyupHandler($event); })("focus", function MatTimepickerDirective_focus_HostBindingHandler() { return ctx.focusHandler(); })("focusout", function MatTimepickerDirective_focusout_HostBindingHandler() { return ctx.focusoutHandler(); });
    } if (rf & 2) {
        ɵngcc0.ɵɵhostProperty("disabled", ctx.disabled)("required", ctx.required);
        ɵngcc0.ɵɵattribute("id", ctx.id)("placeholder", ctx.placeholder)("readonly", ctx.readonly || null)("aria-invalid", ctx.errorState)("aria-required", ctx.required.toString())("aria-describedby", ctx.describedBy);
        ɵngcc0.ɵɵclassProp("mat-input-server", ctx._isServer)("floating", ctx.shouldLabelFloat);
    } }, inputs: { okLabel: "okLabel", cancelLabel: "cancelLabel", mode: "mode", color: "color", disableDialogOpenOnClick: "disableDialogOpenOnClick", strict: "strict", id: "id", errorStateMatcher: "errorStateMatcher", disabled: "disabled", readonly: "readonly", required: "required", placeholder: "placeholder", value: "value", minDate: "minDate", maxDate: "maxDate" }, outputs: { timeChange: "timeChange", invalidInput: "invalidInput" }, exportAs: ["matTimepicker"], features: [ɵngcc0.ɵɵProvidersFeature([
            { provide: MatFormFieldControl, useExisting: MatTimepickerDirective_1 }
        ]), ɵngcc0.ɵɵNgOnChangesFeature()] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatTimepickerDirective, [{
        type: Directive,
        args: [{
                selector: 'input[matTimepicker]',
                providers: [
                    { provide: MatFormFieldControl, useExisting: MatTimepickerDirective_1 }
                ],
                // tslint:disable-next-line:no-host-metadata-property
                host: {
                    /**
                     * @breaking-change 8.0.0 remove .mat-form-field-autofill-control in favor of AutofillMonitor.
                     */
                    // tslint:disable-next-line:object-literal-key-quotes
                    'class': 'mat-input-element mat-form-field-autofill-control',
                    '[class.mat-input-server]': '_isServer',
                    // Native input properties that are overwritten by Angular inputs need to be synced with
                    // the native input element. Otherwise property bindings for those don't work.
                    '[attr.id]': 'id',
                    '[attr.placeholder]': 'placeholder',
                    '[disabled]': 'disabled',
                    '[required]': 'required',
                    '[attr.readonly]': 'readonly || null',
                    '[attr.aria-invalid]': 'errorState',
                    '[attr.aria-required]': 'required.toString()'
                },
                exportAs: 'matTimepicker'
            }]
    }], function () { return [{ type: ɵngcc5.NgControl, decorators: [{
                type: Optional
            }, {
                type: Self
            }] }, { type: ɵngcc1.MatDialog }, { type: ɵngcc0.Renderer2 }, { type: ɵngcc0.NgZone }, { type: ɵngcc6.FocusMonitor }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.NgZone }, { type: ɵngcc7.Platform }, { type: ɵngcc5.NgForm, decorators: [{
                type: Optional
            }] }, { type: ɵngcc8.MatFormField, decorators: [{
                type: Optional
            }] }, { type: ɵngcc5.FormGroupDirective, decorators: [{
                type: Optional
            }] }, { type: ɵngcc9.ErrorStateMatcher }]; }, { describedBy: [{
            type: HostBinding,
            args: ['attr.aria-describedby']
        }], okLabel: [{
            type: Input
        }], cancelLabel: [{
            type: Input
        }], mode: [{
            type: Input
        }], color: [{
            type: Input
        }], disableDialogOpenOnClick: [{
            type: Input
        }], strict: [{
            type: Input
        }], timeChange: [{
            type: Output
        }], invalidInput: [{
            type: Output
        }], id: [{
            type: Input
        }], errorStateMatcher: [{
            type: Input
        }], disabled: [{
            type: Input
        }], readonly: [{
            type: Input
        }], shouldLabelFloat: [{
            type: HostBinding,
            args: ['class.floating']
        }], required: [{
            type: Input
        }], placeholder: [{
            type: Input
        }], value: [{
            type: Input
        }], inputHandler: [{
            type: HostListener,
            args: ['input']
        }], keydownHandler: [{
            type: HostListener,
            args: ['keydown', ['$event']]
        }], keyupHandler: [{
            type: HostListener,
            args: ['keyup', ['$event']]
        }], focusHandler: [{
            type: HostListener,
            args: ['focus']
        }], focusoutHandler: [{
            type: HostListener,
            args: ['focusout']
        }], minDate: [{
            type: Input
        }], maxDate: [{
            type: Input
        }] }); })();
    return MatTimepickerDirective;
}());

var ClockComponent = /** @class */ (function () {
    function ClockComponent() {
        var _this = this;
        this.color = 'primary';
        this.changeEvent = new EventEmitter();
        this.unavailableSelection = new EventEmitter();
        this.invalidMeridiem = new EventEmitter();
        this.invalidSelection = new EventEmitter();
        this.clearInvalidMeridiem = new EventEmitter();
        this.allowed12HourMap = null;
        this.allowed24HourMap = null;
        this.isFormattedValueAllowed = true;
        this.meridiem = null;
        this.touching = false;
        this.numbers = [];
        this.secondaryNumbers = [];
        this.minuteDots = [];
        this.invalidMeridiemEmitted = true;
        this.handleTouchMove = function (e) {
            e.preventDefault(); // prevent scrolling behind the clock on iOS
            var rect = e.target.getBoundingClientRect();
            _this.movePointer(e.changedTouches[0].clientX - rect.left, e.changedTouches[0].clientY - rect.top);
        };
    }
    ClockComponent.prototype.initIsAllowedFn = function () {
        if (!this.allowed12HourMap && !this.allowed24HourMap) {
            return;
        }
        this.isAvailableFn = getIsAvailabeFn(this.allowed12HourMap, this.allowed24HourMap, this.mode);
    };
    ClockComponent.prototype.isAvailable = function (value) {
        return this.isAvailableFn ? this.isAvailableFn(value, this.viewType, this.isPm, this.formattedHours) : true;
    };
    ClockComponent.prototype.ngOnChanges = function (simpleChanges) {
        var _this = this;
        if (simpleChanges.allowed12HourMap ||
            simpleChanges.allowed24HourMap ||
            (simpleChanges.mode && !simpleChanges.mode.firstChange)) {
            this.initIsAllowedFn();
        }
        this.calculateAngule();
        this.setNumbers();
        this.meridiem = this.isPm ? 'PM' : 'AM';
        if (simpleChanges.formattedValue && (this.allowed12HourMap || this.allowed24HourMap)) {
            this.isFormattedValueAllowed = this.isAvailable(this.formattedValue);
        }
        var isSelectedTimeAvailable = (this.isAvailableFn) ?
            this.isAvailableFn(this.formattedValue, 'minutes', this.isPm, this.formattedHours) : true;
        if (this.mode === '24h' && this.viewType === 'minutes' && this.isAvailableFn) {
            var areMinitesAvailable = this.isAvailableFn(this.minutes, 'minutes', this.isPm, this.formattedHours);
            if (!areMinitesAvailable) {
                if (this.minDate && this.minDate.getMinutes() > this.minutes) {
                    setTimeout(function () { _this.changeEvent.emit({ value: _this.minDate.getMinutes(), type: 'minutes' }); });
                }
                else {
                    setTimeout(function () { _this.changeEvent.emit({ value: _this.maxDate.getMinutes(), type: 'minutes' }); });
                }
            }
        }
        if (isSelectedTimeAvailable && this.invalidMeridiemEmitted) {
            this.clearInvalidMeridiem.emit();
            this.invalidMeridiemEmitted = false;
        }
        this.invalidSelection.emit(!isSelectedTimeAvailable);
    };
    ClockComponent.prototype.calculateAngule = function () {
        this.angle = this.getPointerAngle(this.formattedValue, this.viewType);
    };
    ClockComponent.prototype.setNumbers = function () {
        var _this = this;
        if (this.viewType === 'hours') {
            if (this.mode === '12h') {
                var meridiem_1 = this.isPm ? 'pm' : 'am';
                var isAllowedFn = this.allowed12HourMap ? function (num) { return _this.allowed12HourMap[meridiem_1][num + 1][0]; } : undefined;
                this.numbers = this.getNumbers(12, { size: 256 }, isAllowedFn);
                this.secondaryNumbers = [];
                this.minuteDots = [];
            }
            else if (this.mode === '24h') {
                var isAllowedFn = this.allowed24HourMap ? function (num) { return _this.allowed24HourMap[num][0]; } : undefined;
                this.numbers = this.getNumbers(12, { size: 256 }, isAllowedFn);
                this.secondaryNumbers = this.getNumbers(12, { size: 256 - 64, start: 13 }, isAllowedFn);
                this.minuteDots = [];
            }
        }
        else {
            var meridiem_2 = this.isPm ? 'pm' : 'am';
            var isAllowedFn_1 = !!this.allowed12HourMap ? function (num) { return _this.allowed12HourMap[meridiem_2][_this.formattedHours][num]; } :
                !!this.allowed24HourMap ? function (num) { return _this.allowed24HourMap[_this.formattedHours][num]; } : undefined;
            this.numbers = this.getNumbers(12, { size: 256, start: 5, step: 5 }, isAllowedFn_1);
            this.minuteDots = this.getNumbers(60, { size: 256, start: 13 }).map(function (digit) {
                if (digit.display <= 59) {
                    digit.allowed = isAllowedFn_1 ? isAllowedFn_1(digit.display) : true;
                    return digit;
                }
                digit.display = digit.display - 60;
                digit.allowed = isAllowedFn_1 ? isAllowedFn_1(digit.display) : true;
                return digit;
            });
            this.secondaryNumbers = [];
        }
    };
    ClockComponent.prototype.disableAnimatedPointer = function () {
        this.touching = true;
    };
    ClockComponent.prototype.enableAnimatedPointer = function () {
        this.touching = false;
    };
    ClockComponent.prototype.handleTouchEnd = function (e) {
        this.handleTouchMove(e);
        this.enableAnimatedPointer();
    };
    ClockComponent.prototype.handleMouseMove = function (e) {
        // MouseEvent.which is deprecated, but MouseEvent.buttons is not supported in Safari
        if (e.buttons === 1 || e.which === 1) {
            var rect = e.target.getBoundingClientRect();
            this.movePointer(e.clientX - rect.left, e.clientY - rect.top);
        }
    };
    ClockComponent.prototype.handleClick = function (e) {
        var rect = e.target.getBoundingClientRect();
        this.movePointer(e.clientX - rect.left, e.clientY - rect.top);
    };
    ClockComponent.prototype.movePointer = function (x, y) {
        var value = this.getPointerValue(x, y, 256);
        if (!this.isAvailable(value)) {
            this.unavailableSelection.emit();
            return;
        }
        if (value !== this.formattedValue) {
            this.changeEvent.emit({ value: value, type: this.viewType !== 'minutes' ? 'hours' : 'minutes' });
            if (this.viewType !== 'minutes') {
                if (!this.isAvailable(value)) {
                    if (this.minDate && this.isAvailable(value)) {
                        this.changeEvent.emit({ value: this.minDate.getMinutes(), type: 'minutes' });
                    }
                    else if (this.maxDate && this.isAvailable(value)) {
                        this.changeEvent.emit({ value: this.maxDate.getMinutes(), type: 'minutes' });
                    }
                }
            }
        }
    };
    ClockComponent.prototype.getNumbers = function (count, _a, isAllowedFn) {
        var size = _a.size, _b = _a.start, start = _b === void 0 ? 1 : _b, _c = _a.step, step = _c === void 0 ? 1 : _c;
        return Array.apply(null, Array(count)).map(function (_, i) { return ({
            display: i * step + start,
            translateX: (size / 2 - 20) * Math.cos(2 * Math.PI * (i - 2) / count),
            translateY: (size / 2 - 20) * Math.sin(2 * Math.PI * (i - 2) / count),
            allowed: isAllowedFn ? isAllowedFn(i) : true
        }); });
    };
    ClockComponent.prototype.getPointerAngle = function (value, mode) {
        if (this.viewType === 'hours') {
            return this.mode === '12h' ? 360 / 12 * (value - 3) : 360 / 12 * (value % 12 - 3);
        }
        return 360 / 60 * (value - 15);
    };
    ClockComponent.prototype.getPointerValue = function (x, y, size) {
        var value;
        var angle = Math.atan2(size / 2 - x, size / 2 - y) / Math.PI * 180;
        if (angle < 0) {
            angle = 360 + angle;
        }
        if (this.viewType === 'hours') {
            if (this.mode === '12h') {
                value = 12 - Math.round(angle * 12 / 360);
                return value === 0 ? 12 : value;
            }
            var radius = Math.sqrt(Math.pow(size / 2 - x, 2) + Math.pow(size / 2 - y, 2));
            value = 12 - Math.round(angle * 12 / 360);
            if (value === 0) {
                value = 12;
            }
            if (radius < size / 2 - 32) {
                value = value === 12 ? 0 : value + 12;
            }
            return value;
        }
        value = Math.round(60 - 60 * angle / 360);
        return value === 60 ? 0 : value;
    };
    __decorate([
        Input()
    ], ClockComponent.prototype, "mode", void 0);
    __decorate([
        Input()
    ], ClockComponent.prototype, "viewType", void 0);
    __decorate([
        Input()
    ], ClockComponent.prototype, "color", void 0);
    __decorate([
        Input()
    ], ClockComponent.prototype, "formattedValue", void 0);
    __decorate([
        Input()
    ], ClockComponent.prototype, "minDate", void 0);
    __decorate([
        Input()
    ], ClockComponent.prototype, "maxDate", void 0);
    __decorate([
        Input()
    ], ClockComponent.prototype, "isPm", void 0);
    __decorate([
        Input()
    ], ClockComponent.prototype, "formattedHours", void 0);
    __decorate([
        Input()
    ], ClockComponent.prototype, "minutes", void 0);
    __decorate([
        Output()
    ], ClockComponent.prototype, "changeEvent", void 0);
    __decorate([
        Output()
    ], ClockComponent.prototype, "unavailableSelection", void 0);
    __decorate([
        Output()
    ], ClockComponent.prototype, "invalidMeridiem", void 0);
    __decorate([
        Output()
    ], ClockComponent.prototype, "invalidSelection", void 0);
    __decorate([
        Output()
    ], ClockComponent.prototype, "clearInvalidMeridiem", void 0);
    __decorate([
        Input()
    ], ClockComponent.prototype, "allowed12HourMap", void 0);
    __decorate([
        Input()
    ], ClockComponent.prototype, "allowed24HourMap", void 0);
ClockComponent.ɵfac = function ClockComponent_Factory(t) { return new (t || ClockComponent)(); };
ClockComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ClockComponent, selectors: [["mat-clock"]], inputs: { color: "color", allowed12HourMap: "allowed12HourMap", allowed24HourMap: "allowed24HourMap", mode: "mode", viewType: "viewType", formattedValue: "formattedValue", minDate: "minDate", maxDate: "maxDate", isPm: "isPm", formattedHours: "formattedHours", minutes: "minutes" }, outputs: { changeEvent: "changeEvent", unavailableSelection: "unavailableSelection", invalidMeridiem: "invalidMeridiem", invalidSelection: "invalidSelection", clearInvalidMeridiem: "clearInvalidMeridiem" }, features: [ɵngcc0.ɵɵNgOnChangesFeature()], decls: 10, vars: 16, consts: [[1, "root"], [1, "circle", 3, "touchmove", "mousemove", "touchstart", "mousedown", "touchend", "mouseup", "click"], [1, "pointer-container", 3, "ngClass"], ["mat-mini-fab", "", 1, "inner-dot", 3, "color"], [1, "pointer", 3, "color"], ["mat-mini-fab", "", 1, "outer-dot", 3, "color", "ngClass"], [4, "ngIf"], [4, "ngFor", "ngForOf"], ["mat-mini-fab", "", 1, "number", "minute-dot", 3, "ngClass"], ["mat-mini-fab", "", 1, "number", 3, "color", "ngClass"], ["hoursTemplate", ""], ["minutesTemplate", ""], [4, "ngIf", "ngIfThen", "ngIfElse"], ["mat-mini-fab", "", 1, "number", "small-number", 3, "color", "ngClass"]], template: function ClockComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵlistener("touchmove", function ClockComponent_Template_div_touchmove_1_listener($event) { return ctx.handleTouchMove($event); })("mousemove", function ClockComponent_Template_div_mousemove_1_listener($event) { return ctx.handleMouseMove($event); })("touchstart", function ClockComponent_Template_div_touchstart_1_listener() { return ctx.disableAnimatedPointer(); })("mousedown", function ClockComponent_Template_div_mousedown_1_listener() { return ctx.disableAnimatedPointer(); })("touchend", function ClockComponent_Template_div_touchend_1_listener($event) { return ctx.handleTouchEnd($event); })("mouseup", function ClockComponent_Template_div_mouseup_1_listener() { return ctx.enableAnimatedPointer(); })("click", function ClockComponent_Template_div_click_1_listener($event) { return ctx.handleClick($event); });
        ɵngcc0.ɵɵelementStart(2, "div", 2);
        ɵngcc0.ɵɵelement(3, "button", 3);
        ɵngcc0.ɵɵelementStart(4, "mat-toolbar", 4);
        ɵngcc0.ɵɵelementStart(5, "button", 5);
        ɵngcc0.ɵɵtemplate(6, ClockComponent_ng_container_6_Template, 2, 0, "ng-container", 6);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(7, ClockComponent_ng_container_7_Template, 3, 7, "ng-container", 7);
        ɵngcc0.ɵɵtemplate(8, ClockComponent_ng_container_8_Template, 7, 10, "ng-container", 7);
        ɵngcc0.ɵɵtemplate(9, ClockComponent_ng_container_9_Template, 3, 8, "ng-container", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵstyleProp("transform", "rotate(" + ctx.angle + "deg)");
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction2(11, _c6, ctx.mode === "24h" && ctx.viewType === "hours" && (ctx.formattedValue === 0 || ctx.formattedValue > 12), !ctx.touching));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("color", ctx.color);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("color", ctx.color);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("color", ctx.color)("ngClass", ɵngcc0.ɵɵpureFunction1(14, _c7, ctx.viewType === "minutes" && ctx.formattedValue % 5 !== 0));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.viewType === "minutes" && ctx.formattedValue % 5 !== 0);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.minuteDots);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.numbers);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.secondaryNumbers);
    } }, directives: [ɵngcc3.NgClass, ɵngcc4.MatButton, ɵngcc2.MatToolbar, ɵngcc3.NgIf, ɵngcc3.NgForOf], styles: [".root[_ngcontent-%COMP%]{width:256px;height:256px;cursor:default}.circle[_ngcontent-%COMP%]{width:256px;height:256px;border-radius:50%;position:relative;background:#ededed;cursor:pointer}.number[_ngcontent-%COMP%]{width:32px;height:32px;left:calc(50% - 16px);top:calc(50% - 16px);position:absolute;text-align:center;line-height:32px;cursor:pointer;font-size:14px;pointer-events:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;display:flex;align-items:center;justify-content:center;flex-direction:column;background-color:transparent;box-shadow:none}.number.disabled[_ngcontent-%COMP%]{color:rgba(1,1,1,.1)}.number[_ngcontent-%COMP%]:not(.selected):not(.disabled){color:rgba(0,0,0,.87)}.number[_ngcontent-%COMP%]:not(.disabled).minute-dot{color:rgba(1,1,1,.7)}.number[_ngcontent-%COMP%]:not(.disabled).minute-dot.selected{color:transparent}.small-number[_ngcontent-%COMP%]{font-size:12px}.small-number[_ngcontent-%COMP%]:not(.selected):not(.disabled){color:rgba(0,0,0,.67)}.pointer-container[_ngcontent-%COMP%]{width:calc(50% - 20px);height:2;position:absolute;left:50%;top:calc(50% - 1px);transform-origin:left center;pointer-events:none}.pointer-container.disabled[_ngcontent-%COMP%]   *[_ngcontent-%COMP%]{background-color:transparent}.pointer[_ngcontent-%COMP%]{height:1px}.animated-pointer[_ngcontent-%COMP%]{transition:.2s ease-out}.small-pointer[_ngcontent-%COMP%]{width:calc(50% - 52px)}.inner-dot[_ngcontent-%COMP%]{position:absolute;top:-3px;left:-4px;width:8px;height:8px;border-radius:50%;box-shadow:none}.outer-dot[_ngcontent-%COMP%]{width:32px;height:32px;position:absolute;right:-16px;box-shadow:none;border-radius:50%;box-sizing:content-box}.outer-dot-odd[_ngcontent-%COMP%]{box-shadow:none;width:32px;height:32px;display:flex;align-items:center;justify-content:center;flex-direction:column}"], changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ClockComponent, [{
        type: Component,
        args: [{
                selector: 'mat-clock',
                template: "<div class=\"root\">\n  <div class=\"circle\" (touchmove)=\"handleTouchMove($event)\" (mousemove)=\"handleMouseMove($event)\"\n    (touchstart)=\"disableAnimatedPointer()\" (mousedown)=\"disableAnimatedPointer()\" (touchend)=\"handleTouchEnd($event)\"\n    (mouseup)=\"enableAnimatedPointer()\" (click)=\"handleClick($event)\">\n    <div class=\"pointer-container\"\n      [ngClass]=\"{ 'small-pointer': mode === '24h' && viewType === 'hours' && (formattedValue === 0 || formattedValue > 12), 'animated-pointer': !touching }\"\n      [style.transform]=\"'rotate(' + angle + 'deg)'\">\n      <button mat-mini-fab [color]=\"color\" class=\"inner-dot\"></button>\n      <mat-toolbar [color]=\"color\" class=\"pointer\">\n        <button mat-mini-fab [color]=\"color\" class=\"outer-dot\"\n          [ngClass]=\" { 'outer-dot-odd': viewType === 'minutes' && formattedValue % 5 !== 0 }\">\n          <ng-container *ngIf=\"viewType === 'minutes' && formattedValue % 5 !== 0\">\u00B7</ng-container>\n        </button>\n      </mat-toolbar>\n    </div>\n    <ng-container *ngFor=\"let digit of minuteDots;\">\n      <button mat-mini-fab class=\"number minute-dot\"\n        [ngClass]=\"{ 'selected': formattedValue === digit.display || (digit.display === 0 && formattedValue === 0), 'disabled': !isAvailable(digit.display === 60 ? 0 : digit.display)}\"\n        [style.transform]=\"'translate(' + digit.translateX + 'px, ' + digit.translateY + 'px)'\">\n        <ng-container *ngIf=\"digit.display % 5 !== 0\">\u00B7</ng-container>\n      </button>\n    </ng-container>\n    <ng-container *ngFor=\"let digit of numbers;\">\n      <button mat-mini-fab [color]=\"color\" class=\"number\"\n        [ngClass]=\"{ 'selected': formattedValue === digit.display || (digit.display === 60 && formattedValue === 0), 'disabled': !isAvailable(digit.display === 60 ? 0 : digit.display)}\"\n        [style.transform]=\"'translate(' + digit.translateX + 'px, ' + digit.translateY + 'px)'\">\n        <ng-template #hoursTemplate>{{ digit.display }}</ng-template>\n        <ng-template #minutesTemplate>{{ digit.display === 60 ? '00' : digit.display }}</ng-template>\n        <ng-container *ngIf=\"viewType === 'minutes' then minutesTemplate else hoursTemplate;\"></ng-container>\n      </button>\n\n    </ng-container>\n    <ng-container *ngFor=\"let digit of secondaryNumbers;\">\n      <button mat-mini-fab [color]=\"color\" class=\"number small-number\"\n        [ngClass]=\"{ 'selected': formattedValue === digit.display || (digit.display === 24 && formattedValue === 0), 'disabled': !isAvailable(digit.display === 24 ? 0 : digit.display)}\"\n        [style.transform]=\"'translate(' + digit.translateX + 'px, ' + digit.translateY + 'px)'\">\n        {{ digit.display === 24 ? '00' : digit.display }}\n      </button>\n\n    </ng-container>\n  </div>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".root{width:256px;height:256px;cursor:default}.circle{width:256px;height:256px;border-radius:50%;position:relative;background:#ededed;cursor:pointer}.number{width:32px;height:32px;left:calc(50% - 16px);top:calc(50% - 16px);position:absolute;text-align:center;line-height:32px;cursor:pointer;font-size:14px;pointer-events:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;display:flex;align-items:center;justify-content:center;flex-direction:column;background-color:transparent;box-shadow:none}.number.disabled{color:rgba(1,1,1,.1)}.number:not(.selected):not(.disabled){color:rgba(0,0,0,.87)}.number:not(.disabled).minute-dot{color:rgba(1,1,1,.7)}.number:not(.disabled).minute-dot.selected{color:transparent}.small-number{font-size:12px}.small-number:not(.selected):not(.disabled){color:rgba(0,0,0,.67)}.pointer-container{width:calc(50% - 20px);height:2;position:absolute;left:50%;top:calc(50% - 1px);transform-origin:left center;pointer-events:none}.pointer-container.disabled *{background-color:transparent}.pointer{height:1px}.animated-pointer{transition:.2s ease-out}.small-pointer{width:calc(50% - 52px)}.inner-dot{position:absolute;top:-3px;left:-4px;width:8px;height:8px;border-radius:50%;box-shadow:none}.outer-dot{width:32px;height:32px;position:absolute;right:-16px;box-shadow:none;border-radius:50%;box-sizing:content-box}.outer-dot-odd{box-shadow:none;width:32px;height:32px;display:flex;align-items:center;justify-content:center;flex-direction:column}"]
            }]
    }], function () { return []; }, { color: [{
            type: Input
        }], changeEvent: [{
            type: Output
        }], unavailableSelection: [{
            type: Output
        }], invalidMeridiem: [{
            type: Output
        }], invalidSelection: [{
            type: Output
        }], clearInvalidMeridiem: [{
            type: Output
        }], allowed12HourMap: [{
            type: Input
        }], allowed24HourMap: [{
            type: Input
        }], mode: [{
            type: Input
        }], viewType: [{
            type: Input
        }], formattedValue: [{
            type: Input
        }], minDate: [{
            type: Input
        }], maxDate: [{
            type: Input
        }], isPm: [{
            type: Input
        }], formattedHours: [{
            type: Input
        }], minutes: [{
            type: Input
        }] }); })();
    return ClockComponent;
}());

var MatTimepickerModule = /** @class */ (function () {
    function MatTimepickerModule() {
    }
MatTimepickerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MatTimepickerModule });
MatTimepickerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MatTimepickerModule_Factory(t) { return new (t || MatTimepickerModule)(); }, imports: [[
            CommonModule,
            MatDialogModule,
            MatButtonModule,
            MatToolbarModule,
            MatIconModule,
            MatInputModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MatTimepickerModule, { declarations: function () { return [ClockComponent,
        MatTimepickerDirective,
        MatTimepickerComponentDialogComponent]; }, imports: function () { return [CommonModule,
        MatDialogModule,
        MatButtonModule,
        MatToolbarModule,
        MatIconModule,
        MatInputModule]; }, exports: function () { return [MatTimepickerDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MatTimepickerModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    ClockComponent,
                    MatTimepickerDirective,
                    MatTimepickerComponentDialogComponent
                ],
                imports: [
                    CommonModule,
                    MatDialogModule,
                    MatButtonModule,
                    MatToolbarModule,
                    MatIconModule,
                    MatInputModule
                ],
                exports: [
                    MatTimepickerDirective
                ],
                entryComponents: [
                    MatTimepickerComponentDialogComponent
                ]
            }]
    }], function () { return []; }, null); })();
    return MatTimepickerModule;
}());

/*
 * Public API Surface of mat-timepicker
 */

/**
 * Generated bundle index. Do not edit.
 */

export { MatTimepickerDirective, MatTimepickerModule, ClockComponent as ɵa, MatTimepickerComponentDialogComponent as ɵb };
